src/hotspot/share/gc/shenandoah/shenandoahGenerationalEvacuationTask.cpp

```diff
@@ -68,16 +69,24 @@ void ShenandoahGenerationalEvacuationTask::work(uint worker_id) {
   if (_concurrent) {
     ShenandoahConcurrentWorkerSession worker_session(worker_id);
     ShenandoahSuspendibleThreadSetJoiner stsj;
-    ShenandoahEvacOOMScope oom_evac_scope;
     do_work();
   } else {
     ShenandoahParallelWorkerSession worker_session(worker_id);
-    ShenandoahEvacOOMScope oom_evac_scope;
     do_work();
   }
 }
 
 void ShenandoahGenerationalEvacuationTask::do_work() {
+  if (_only_promote_regions) {
+    // No allocations will be made, do not enter oom-during-evac protocol.
+    do_evacuations();
```

Comment:
It really looks to me that promotion-in-place should be in a separate task? Or the name of the task should reflect it also does promotions? Or at least the "real evacuation" paths in `do_evacuations` be asserted with `!_only_promote_regions`?

Reply:
I didn't go so far as to split them into separate tasks, but I did factor the different paths into separate methods. Also added an assert that the collection set is empty when we are only promoting regions in place.